### PR TITLE
cgroups: Use only pod cgroup

### PR DESCRIPTION
**Pull request by jcvenegas**
- config add option to use only pod cgroup (`SandboxCgroupOnly`)
- Get the sandbox Cgroup path based on cgroupsPath from `Sandbox` Container 
- Join the Sandbox Cgroup before create any process
- All the kata processes are inside the Sandbox Cgroup
- The cgroup is not restricted by the runtime but the caller  is free to limit it. 

Depends-on: github.com/kata-containers/tests#1824

Fixes: #1879
Signed-off-by: Jose Carlos Venegas Munoz <jose.carlos.venegas.munoz@intel.com>

### CODE REVIEW

File: virtcontainers/container.go [hunk 1]
```diff
@@ -1435,19 +1436,46 @@ func (c *Container) newCgroups() (err error) {
...
+		if cgroupPath == "" {
+			c.Logger().Warn("Sandbox container with not cgroup path provided using / as default")
+			spec.Linux.CgroupsPath = "/"
+		}
+		c.sandbox.state.CgroupPath = filepath.Dir(cgroupPath)
+		c.Logger().WithFields(
+			logrus.Fields{
+				"container-cgroup": spec.Linux.CgroupsPath,
+				"sandbox-cgroup":   c.sandbox.state.CgroupPath,
+			}).Info("DEBUG: PodSandbox type, using its cgroup parent cgroup path")
```
Comment (jcvenegas) [inline virtcontainers/container.go]: Nit: Remove `DEBUG` adn use .Debug()

File: virtcontainers/sandbox.go [hunk 2]
```diff
@@ -1049,15 +1054,39 @@ func (s *Sandbox) addContainer(c *Container) error {
...
+		// Use the same cgroup of the container sandbox as the sandbox cgroup
 		s.state.CgroupPath = c.state.CgroupPath
-		if !s.supportNewStore() {
-			return s.store.Store(store.State, s.state)
-		}
+		return nil
+	}
+
+	// Use the parent cgroup of the container sandbox as the sandbox cgroup
+	s.state.CgroupPath = filepath.Dir(c.state.CgroupPath) + "/kata-sandbox/"
```
Comment (jcvenegas) [inline virtcontainers/sandbox.go]: We are creating a cgroup, we should should remove it on delete (even if the caller will delete all the parent cgroup)
If we use a common name al the pods will use the same cgroup if the cgroup parent was not set just for the pod, e.g. docker. So probably need to add an additional sufix (container.id) or just drop kata in the parent cgroup instead  of create one.

**Comment by jcvenegas** [comment]
All passed with the flag not enabled. So all good in the default/current way to do cgroups.

Re-ran with  the flag enabled, 

fedora failed with (this job does not run docker)
```
15:26:11 ok 24 ctr oom # skip This is not working (Issue https://github.com/kata-containers/tests/issues/714)
15:26:29 not ok 25 ctr /etc/resolv.conf rw/ro mode
15:26:29 # (in test file ctr.bats, line 882)
15:26:29 #   `[ "$status" -eq 0 ]' failed
```
Restarting to get if is non related error.

Ubuntu job failed with (this job does not run docker)
```
16:30:11 not ok 33 ctr /etc/resolv.conf rw/ro mode
16:30:11 # (in test file ctr.bats, line 1217)
16:30:11 #   `[ "$status" -eq 0 ]' failed
16:30:11 # time="2019-07-16T21:29:09Z" level=error msg="File descriptor 3 (pipe:[156253]) leaked on pvdisplay invocation. Parent PID 109987: /tmp/jenkins/workspace/kata-con
16:30:11 # File descriptor 6 (/dev/mapper/control) leaked on pvdisplay invocation. Parent PID 109987: /tmp/jenkins/workspace/kata-con
16:30:11 #   Failed to find physical volume "/dev/sdb".
16:30:11 # " error="exit status 5" 
```

The rest of jobs that failed with 
```
15:23:10 [1] • Failure [13.542 seconds]
15:23:10 [1] Checking CPU cgroups in the host
15:23:10 [1] /tmp/jenkins/workspace/kata-containers-runtime-ubuntu-nemu/go/src/github.com/kata-containers/tests/integration/docker/cgroups_test.go:110
15:23:10 [1]   checking whether cgroups can be deleted
15:23:10 [1]   /tmp/jenkins/workspace/kata-containers-runtime-ubuntu-nemu/go/src/github.com/kata-containers/tests/integration/docker/cgroups_test.go:130
15:23:10 [1]     with a running process
15:23:10 [1]     /tmp/jenkins/workspace/kata-containers-runtime-ubuntu-nemu/go/src/github.com/kata-containers/tests/integration/docker/cgroups_test.go:131
15:23:10 [1]       should be deleted [It]
15:23:10 [1]       /tmp/jenkins/workspace/kata-containers-runtime-ubuntu-nemu/go/src/github.com/kata-containers/tests/integration/docker/cgroups_test.go:132
15:23:10 [1] 
15:23:10 [1]       Expected
15:23:10 [1]           <string>: /sys/fs/cgroup/cpu/docker/kata_ae38701e4bc415ee6876f4dba2d1841949f0528f4bbc336ac7526cd4634445e8
15:23:10 [1]       to be a directory: stat /sys/fs/cgroup/cpu/docker/kata_ae38701e4bc415ee6876f4dba2d1841949f0528f4bbc336ac7526cd4634445e8: no such file or directory
```
Expected as those run docker test

File: virtcontainers/container.go [hunk 3]
```diff
@@ -1406,8 +1408,12 @@ func (c *Container) detachDevices() error {
 	return nil
 }
 
-// creates a new cgroup and return the cgroups path
+// creates a new cgroup for a container
+// If the container is a sandbox type
+// set the sandbox cgroup to the container cgroup path
+// Add the shim to the container cgroup
 func (c *Container) newCgroups() (err error) {
+
```
Comment (jcvenegas) [inline virtcontainers/container.go]: Nit: blank

File: cli/config/configuration-acrn.toml.in [hunk 4]
```diff
@@ -224,6 +224,12 @@ disable_guest_seccomp=@DEFDISABLEGUESTSECCOMP@
 # If you are using docker, `disable_new_netns` only works with `docker run --net=none`
 # (default: false)
 #disable_new_netns = true
+#
+# if enable, the runtime use the parent cgroup of a container PodSandbox.  This
+# should be enabled for users where the caller setup the parent cgroup of the
+# containers running in a sandbox so all the resouces of the kata container run
+# in the same cgroup and performance isolation its more accurate.
+sandbox_cgroup_only=false
```
Comment (devimc) [inline cli/config/configuration-acrn.toml.in]: please change to `true`, comment it and set to `false` in the code, this way would be easier for the user enable it.

File: cli/config/configuration-fc.toml.in [hunk 5]
```diff
@@ -315,6 +315,12 @@ disable_guest_seccomp=@DEFDISABLEGUESTSECCOMP@
 # (default: false)
 #disable_new_netns = true
 
+# if enable, the runtime use the parent cgroup of a container PodSandbox.  This
+# should be enabled for users where the caller setup the parent cgroup of the
+# containers running in a sandbox so all the resouces of the kata container run
+# in the same cgroup and performance isolation its more accurate.
+sandbox_cgroup_only=false
```
Comment (devimc) [inline cli/config/configuration-fc.toml.in]: please change to true, comment it and set to false in the code, this way would be easier for the user enable it.

File: cli/config/configuration-nemu.toml.in [hunk 6]
```diff
@@ -389,6 +389,13 @@ disable_guest_seccomp=@DEFDISABLEGUESTSECCOMP@
 # (default: false)
 #disable_new_netns = true
 
+# if enable, the runtime use the parent cgroup of a container PodSandbox.  This
+# should be enabled for users where the caller setup the parent cgroup of the
+# containers running in a sandbox so all the resouces of the kata container run
+# in the same cgroup and performance isolation its more accurate.
+sandbox_cgroup_only=false
```
Comment (devimc) [inline cli/config/configuration-nemu.toml.in]: same here

File: cli/config/configuration-qemu.toml.in [hunk 7]
```diff
@@ -396,6 +396,12 @@ disable_guest_seccomp=@DEFDISABLEGUESTSECCOMP@
 # (default: false)
 #disable_new_netns = true
 
+# if enable, the runtime use the parent cgroup of a container PodSandbox.  This
+# should be enabled for users where the caller setup the parent cgroup of the
+# containers running in a sandbox so all the resouces of the kata container run
+# in the same cgroup and performance isolation its more accurate.
+sandbox_cgroup_only=false
```
Comment (devimc) [inline cli/config/configuration-qemu.toml.in]: and here

File: virtcontainers/cgroups.go [hunk 8]
```diff
@@ -145,22 +145,34 @@ func parentCgroup(hierarchy cgroups.Hierarchy, path string) (cgroups.Cgroup, err
 }
 
 func (s *Sandbox) updateCgroups() error {
+
+	if s.config != nil && s.config.SandboxCgroupOnly {
+		s.Logger().Debug("SandboxCgroupOnly, runtime previously join to sandbox cgroup - nothing to do")
```
Comment (devimc) [inline virtcontainers/cgroups.go]: how about using `WithField` ?

File: virtcontainers/cgroups.go [hunk 9]
```diff
@@ -145,45 +145,68 @@ func parentCgroup(hierarchy cgroups.Hierarchy, path string) (cgroups.Cgroup, err
...
+
 	if s.state.CgroupPath == "" {
 		s.Logger().Warn("sandbox's cgroup won't be updated: cgroup path is empty")
 		return nil
 	}
 
-	cgroup, err := cgroupsLoadFunc(V1Constraints, cgroups.StaticPath(s.state.CgroupPath))
+	cgroupHierarchy := V1Constraints
+
+	if s.config == nil {
```
Comment (jodh-intel) [inline virtcontainers/cgroups.go]: Can you move this test to the start of the function? It seems a bit odd doing it here as `s.config` doesn't (appear) to be used below this test?!? Moving it up would also allow you to simplify the `s.config.SandboxCgroupOnly` test above.

File: virtcontainers/cgroups.go [hunk 10]
```diff
@@ -145,45 +145,68 @@ func parentCgroup(hierarchy cgroups.Hierarchy, path string) (cgroups.Cgroup, err
...
 
-	path := cgroupNoConstraintsPath(s.state.CgroupPath)
-	s.Logger().WithField("path", path).Debug("Deleting no constraints cgroup")
-	noConstraintsCgroup, err := cgroupsLoadFunc(V1NoConstraints, cgroups.StaticPath(path))
+	var path string
+	cgroupTypes := cgroups.V1
+	if s.config.SandboxCgroupOnly {
+		path = s.state.CgroupPath
+		s.Logger().WithField("path", path).Debug("Deleting sandbox cgroup")
+
```
Comment (jodh-intel) [inline virtcontainers/cgroups.go]: Nit: blank line.

File: virtcontainers/cgroups.go [hunk 11]
```diff
@@ -145,45 +145,68 @@ func parentCgroup(hierarchy cgroups.Hierarchy, path string) (cgroups.Cgroup, err
...
+	} else {
+		path = cgroupNoConstraintsPath(s.state.CgroupPath)
+		s.Logger().WithField("path", path).Debug("Deleting no constraints sandbox cgroup")
+		cgroupTypes = V1NoConstraints
+	}
+
+	cgroup, err := cgroupsLoadFunc(cgroupTypes, cgroups.StaticPath(path))
 	if err == cgroups.ErrCgroupDeleted {
-		// cgroup already deleted
+		s.Logger().Debug("cgroup already deleted: ", path)
```
Comment (jodh-intel) [inline virtcontainers/cgroups.go]:
```go
s.Logger().WithField("path", path).Debug("cgroup already deleted")
```

File: virtcontainers/container.go [hunk 12]
```diff
@@ -1488,27 +1498,32 @@ func (c *Container) deleteCgroups() error {
...
-	}
-
 	// store new resources
 	c.config.Resources = r
 	if err := c.storeContainer(); err != nil {
 		return err
 	}
 
+	if c.sandbox.config.SandboxCgroupOnly {
+		c.Logger().Debug("using pod cgroup: not need to update container cgroup for container in host")
```
Comment (jodh-intel) [inline virtcontainers/container.go]: Nit: s/not need/no need/

File: virtcontainers/sandbox.go [hunk 13]
```diff
@@ -1073,13 +1079,59 @@ func (s *Sandbox) addContainer(c *Container) error {
...
+
+		validContainerCgroup := utils.ValidCgroupPath(spec.Linux.CgroupsPath)
+		if !s.config.SandboxCgroupOnly {
+			// Use the same cgroup of the container sandbox as the sandbox cgroup
+			s.state.CgroupPath = validContainerCgroup
+			return nil
 		}
+
+		// Use the parent cgroup of the container sandbox as the sandbox cgroup
+		s.state.CgroupPath = filepath.Dir(validContainerCgroup) + "/kata-sandbox-" + cConfig.ID
```
Comment (jodh-intel) [inline virtcontainers/sandbox.go]: Rather than using `+`, it would be better to use `filepath.Join()`. Also, that fixed string could be a top-level `const` maybe.

File: virtcontainers/sandbox.go [hunk 14]
```diff
@@ -1874,3 +1930,27 @@ func (s *Sandbox) calculateSandboxCPUs() uint32 {
...
+
+	cgroup, err := cgroupsLoadFunc(cgroups.V1, cgroups.StaticPath(s.state.CgroupPath))
+	if err != nil {
+		return fmt.Errorf("Could not load sandbox cgroup in %v: %v", s.state.CgroupPath, err)
+	}
+
+	s.Logger().WithField("cgroup:", s.state.CgroupPath).Debug("joining to sandbox cgroup")
+	runtimePid := os.Getpid()
+
+	if runtimePid <= 0 {
```
Comment (jodh-intel) [inline virtcontainers/sandbox.go]: Nit: technically, `getpid()` cannot fail.

File: cli/config/configuration-fc.toml.in [hunk 15]
```diff
@@ -315,6 +315,12 @@ disable_guest_seccomp=@DEFDISABLEGUESTSECCOMP@
 # (default: false)
 #disable_new_netns = true
 
+# if enable, the runtime use the parent cgroup of a container PodSandbox.  This
```
Comment (jodh-intel) [inline cli/config/configuration-fc.toml.in]: We *really* need to create a bunch of TOML fragments real soon now (TM) to avoid having to update all these config files ;(

File: cli/config/configuration-acrn.toml.in [hunk 16]
```diff
@@ -224,6 +224,12 @@ disable_guest_seccomp=@DEFDISABLEGUESTSECCOMP@
 # If you are using docker, `disable_new_netns` only works with `docker run --net=none`
 # (default: false)
 #disable_new_netns = true
+#
+# if enable, the runtime use the parent cgroup of a container PodSandbox.  This
```
Comment (jodh-intel) [inline cli/config/configuration-acrn.toml.in]: > If enabled, the runtime uses the parent "PodSandbox" cgroup.

I'm afraid I got lost with the second sentence - can you break it into two sentences maybe to make it easier to understand?

**Comment by jodh-intel** [comment]
@jcvenegas - does this PR replace https://github.com/kata-containers/runtime/pull/1431?

File: cli/config/configuration-acrn.toml.in (hunk 16 above)
Comment (jcvenegas) [reply cli/config/configuration-acrn.toml.in]: Hey @jodh-intel  I have tried to re-work the sentences, let me know if that works. I still feel is not good enough the main thing is that this option works based in container types. Specially `PodSanbox`  I wanted to point to some documentation but seems that we dont have documentation about it.

File: cli/config/configuration-acrn.toml.in [hunk 17]
```diff
@@ -225,6 +225,14 @@ disable_guest_seccomp=@DEFDISABLEGUESTSECCOMP@
 # (default: false)
 #disable_new_netns = true
 
+# if enable, the runtime will add all the kata processes inside one dedicated cgroup.
```
Comment (egernst) [inline cli/config/configuration-acrn.toml.in]: "If enabled,"

File: virtcontainers/api.go [hunk 18]
```diff
@@ -77,6 +77,17 @@ func createSandboxFromConfig(ctx context.Context, sandboxConfig SandboxConfig, f
 		return nil, err
 	}
 
+	// Move runtime to sandbox cgroup so all process are created there.
+	if err := s.retrieveCgroupPath(); err != nil {
+		return nil, err
+	}
+	// Join to sandbox cgroup before create any asset so all will be created there
```
Comment (egernst) [inline virtcontainers/api.go]: 'Join the sandbox cgroup before creating any processes, so they are all created within that cgroup'

^ -- does that make more sense?

File: virtcontainers/cgroups.go [hunk 19]
```diff
@@ -145,45 +146,68 @@ func parentCgroup(hierarchy cgroups.Hierarchy, path string) (cgroups.Cgroup, err
...
+	if s.config == nil {
+		s.Logger().Warn("sandbox.config is nil")
+		return nil
+	}
+
+	if s.config.SandboxCgroupOnly {
+		s.Logger().WithField(
+			"SandboxCgroupOnly",
+			s.config.SandboxCgroupOnly).
+			Debug("runtime previously join to sandbox cgroup - nothing to do")
```
Comment (egernst) [inline virtcontainers/cgroups.go]: "runtime previously joined sandbox cgroup - nothing to do"

File: virtcontainers/container.go [hunk 20]
```diff
@@ -1488,27 +1498,32 @@ func (c *Container) deleteCgroups() error {
...
-		return fmt.Errorf("Could not update cgroup %v: %v", c.state.CgroupPath, err)
-	}
-
 	// store new resources
 	c.config.Resources = r
 	if err := c.storeContainer(); err != nil {
 		return err
 	}
 
+	if c.sandbox.config.SandboxCgroupOnly {
```
Comment (egernst) [inline virtcontainers/container.go]: should this clause be at top of function?

File: virtcontainers/container.go [hunk 21]
```diff
@@ -1451,6 +1460,7 @@ func (c *Container) newCgroups() (err error) {
 	}
 
 	return nil
+
```
Comment (jodh-intel) [inline virtcontainers/container.go]: ?

File: virtcontainers/cgroups.go [hunk 22]
```diff
@@ -145,45 +146,68 @@ func parentCgroup(hierarchy cgroups.Hierarchy, path string) (cgroups.Cgroup, err
 }
 
 func (s *Sandbox) updateCgroups() error {
+
```
Comment (jodh-intel) [inline virtcontainers/cgroups.go]: Nit: extra line.

File: virtcontainers/sandbox.go [hunk 23]
```diff
@@ -1073,14 +1079,69 @@ func (s *Sandbox) addContainer(c *Container) error {
...
+
+		} else {
+			// Use the same cgroup of the container sandbox as the sandbox cgroup
+			kataContainergroupPath, err := renameCgroupPath(validContainerCgroup)
+			if err != nil {
+				return err
+			}
+			s.state.CgroupPath = kataContainergroupPath
+		}
+		if ann[annotations.ContainerTypeKey] == string(PodSandbox) {
```
Comment (jodh-intel) [inline virtcontainers/sandbox.go]: Nit: missing blank line.

File: virtcontainers/sandbox.go [hunk 24]
```diff
@@ -1073,14 +1079,69 @@ func (s *Sandbox) addContainer(c *Container) error {
...
+			break
+		}
+	}
+	if s.state.CgroupPath == "" {
```
Comment (jodh-intel) [inline virtcontainers/sandbox.go]: Needs a blank line.

File: virtcontainers/sandbox.go [hunk 25]
```diff
@@ -1073,14 +1079,69 @@ func (s *Sandbox) addContainer(c *Container) error {
...
+		return nil
 	}
 
+	_, err := cgroupsNewFunc(cgroups.V1, cgroups.StaticPath(s.state.CgroupPath), &specs.LinuxResources{})
+	if err != nil {
+		return fmt.Errorf("Could not create sandbox cgroup in %v: %v", s.state.CgroupPath, err)
+	}
+	return nil
+}
+func (s *Sandbox) addFetchedContainer(c *Container) error {
```
Comment (jodh-intel) [inline virtcontainers/sandbox.go]: Needs a blank line.

File: virtcontainers/sandbox.go [hunk 26]
```diff
@@ -1073,14 +1079,69 @@ func (s *Sandbox) addContainer(c *Container) error {
...
+	if _, ok := s.containers[c.id]; ok {
+		return fmt.Errorf("Duplicated container: %s", c.id)
+	}
+	s.containers[c.id] = c
```
Comment (jodh-intel) [inline virtcontainers/sandbox.go]: Missing blank line.

File: virtcontainers/sandbox.go [hunk 27]
```diff
@@ -1874,3 +1939,23 @@ func (s *Sandbox) calculateSandboxCPUs() uint32 {
 func (s *Sandbox) GetHypervisorType() string {
 	return string(s.config.HypervisorType)
 }
+
+func (s *Sandbox) joinSandboxCgroup() error {
+
+	if s.state.CgroupPath == "" {
+		s.Logger().Debug("sandbox cgroup emtpy, not joining")
```
Comment (jodh-intel) [inline virtcontainers/sandbox.go]: - Typo: "empty".
- Missing `return nil`?

File: virtcontainers/sandbox.go [hunk 28]
```diff
@@ -1874,3 +1939,23 @@ func (s *Sandbox) calculateSandboxCPUs() uint32 {
...
+		return fmt.Errorf("Could not load sandbox cgroup in %v: %v", s.state.CgroupPath, err)
+	}
+
+	s.Logger().WithField("cgroup:", s.state.CgroupPath).Debug("joining to sandbox cgroup")
+	runtimePid := os.Getpid()
+
+	if err := cgroup.Add(cgroups.Process{Pid: runtimePid}); err != nil {
+		return fmt.Errorf("Could not add runtime PID %d to sandbox cgroup:  %v", runtimePid, err)
+	}
+	return nil
```
Comment (jodh-intel) [inline virtcontainers/sandbox.go]: Missing blank line.

File: virtcontainers/container.go (hunk 20 above)
Comment (jcvenegas) [reply virtcontainers/container.go]: Added a comment why is needed here and not in top of function.

File: virtcontainers/cgroups.go [hunk 29]
```diff
@@ -192,28 +215,28 @@ func (s *Sandbox) deleteCgroups() error {
...
 		// Don't fail, cgroup can be deleted
 		s.Logger().WithError(err).Warn("Could not move process from no constraints to parent cgroup")
 	}
-
-	return noConstraintsCgroup.Delete()
+	return cgroup.Delete()
 }
 
-func (s *Sandbox) constrainHypervisor(cgroup cgroups.Cgroup) error {
+func (s *Sandbox) constrainHypervisorVCPUs(cgroup cgroups.Cgroup) error {
```
Comment (egernst) [inline virtcontainers/cgroups.go]: I understand that we are clarifying what the function does by updating the name of it, however, this probably should be in a separate commit, since it isn't really about the cgroup change, right?  Also, /me wonders if we should be more diligent about adding comments for functions, even if they aren't exported.

File: virtcontainers/cgroups.go [hunk 30]
```diff
@@ -145,45 +146,67 @@ func parentCgroup(hierarchy cgroups.Hierarchy, path string) (cgroups.Cgroup, err
...
+		return nil
+	}
+
 	if s.state.CgroupPath == "" {
 		s.Logger().Warn("sandbox's cgroup won't be updated: cgroup path is empty")
 		return nil
 	}
 
-	cgroup, err := cgroupsLoadFunc(V1Constraints, cgroups.StaticPath(s.state.CgroupPath))
+	cgroupHierarchy := V1Constraints
```
Comment (egernst) [inline virtcontainers/cgroups.go]: why is this variable added?

File: virtcontainers/cgroups.go [hunk 31]
```diff
@@ -145,45 +146,67 @@ func parentCgroup(hierarchy cgroups.Hierarchy, path string) (cgroups.Cgroup, err
...
 
 func (s *Sandbox) updateCgroups() error {
+	resources, err := s.resources()
+	if err != nil {
+		return err
+	}
+
+	if s.config == nil {
+		s.Logger().Warn("sandbox.config is nil")
+		return nil
```
Comment (egernst) [inline virtcontainers/cgroups.go]: Should this be an error?  When would this happen in which it is not an error?

File: virtcontainers/cgroups_test.go [hunk 32]
```diff
@@ -134,6 +134,7 @@ func TestUpdateCgroups(t *testing.T) {
 
 	// path doesn't exist
 	s.state.CgroupPath = "/abc/123/rgb"
+	s.config = &SandboxConfig{}
```
Comment (egernst) [inline virtcontainers/cgroups_test.go]: Was this the only of failure for s.config :) ? 

Is it feasible to add a unit test to cover the case we are adding here, podCgroupOnly?

File: virtcontainers/container.go [hunk 33]
```diff
@@ -892,7 +891,10 @@ func (c *Container) delete() error {
 		return err
 	}
 
-	if err := c.deleteCgroups(); err != nil {
+	// Kata does not own the Pod Cgroup
```
Comment (egernst) [inline virtcontainers/container.go]: should this comment be inside the if clause?

File: virtcontainers/container.go (hunk 20 above)
Comment (egernst) [reply virtcontainers/container.go]: Got it -- makes sense!

**Comment by jcvenegas** [comment]
@jodh-intel @egernst  I refactor and split changes in commits let me know if that helps with the review.

- The first 6 commits  cgroups refactor (those could be part of a new PR but because are related with this changes I would like to keep them together)

File: virtcontainers/container.go [hunk 34]
```diff
@@ -1490,6 +1496,11 @@ func (c *Container) cgroupsDelete() error {
 
 // updates container cgroups in the host
 func (c *Container) cgroupsUpdate(resources specs.LinuxResources) error {
+
+	if c.state.CgroupPath == "" {
+		c.Logger().Warn("container has not cgroup path nothing to update")
+		return nil
```
Comment (egernst) [inline virtcontainers/container.go]: Should these be warnings? Seems info or debug maybe?
Suggested comment: “container does not have cgroup on host: nothing to update”?

File: virtcontainers/container.go [hunk 35]
```diff
@@ -1455,6 +1455,12 @@ func (c *Container) cgroupsCreate() (err error) {
 
 // deletes container cgroups in the host
 func (c *Container) cgroupsDelete() error {
+
+	if c.state.CgroupPath == "" {
+		c.Logger().Warn("container has not cgroup path nothing to delete")
```
Comment (egernst) [inline virtcontainers/container.go]: Same as below...

File: virtcontainers/container.go [hunk 36]
```diff
@@ -1503,7 +1503,7 @@ func (c *Container) cgroupsUpdate(resources specs.LinuxResources) error {
 
 	// update cgroup
 	if err := cgroup.Update(&r); err != nil {
```
Comment (egernst) [inline virtcontainers/container.go]: s/cgroups/cgroup in error message here, and below.

File: virtcontainers/sandbox.go [hunk 37]
```diff
@@ -2065,3 +2067,64 @@ func (s *Sandbox) cpuResources() *specs.LinuxCPU {
...
+	_, err := cgroupsNewFunc(cgroups.V1, cgroups.StaticPath(s.state.CgroupPath), &specs.LinuxResources{})
+	if err != nil {
+		return fmt.Errorf("Could not create sandbox cgroup in %v: %v", s.state.CgroupPath, err)
+
+	}
+
+	return nil
+}
+
+// Join to the Sandbox Cgroup, all the process the runtime creates after join  are part of
```
[the remark this answers is not in the document]
Comment (egernst) [reply virtcontainers/sandbox.go]: I’m a bit confused. Should we just utilize the parent cgroup, or create a new one under the parent cgroup? If we do static CPU management I suppose this *may* be necessary, but am not sure.

I could be misunderstanding, but seems we are creating a new one; so we need to account for coy shares, etc?

**Comment by egernst** [comment]
I’m a bit confused. Should we just utilize the parent cgroup, or create a new one under the parent cgroup? If we do static CPU management I suppose this *may* be necessary, but am not sure really. 

I could be misunderstanding, but seems we are creating a new one; so we need to account for cpu shares, etc? This may be more clear in your gist; can we make that a PR to docs repo now and add a depends for this?

Nice work on the cleanup. I agree that unit tests for at least the added functions is a good idea.

File: virtcontainers/api.go [hunk 38]
```diff
@@ -75,6 +75,18 @@ func createSandboxFromConfig(ctx context.Context, sandboxConfig SandboxConfig, f
 		return nil, err
 	}
 
+	// Move runtime to sandbox cgroup so all process are created there.
+	if s.config.SandboxCgroupOnly {
+		if err := s.SetupSandboxCgroupOnly(); err != nil {
```
Comment (egernst) [inline virtcontainers/api.go]: This isn't an exported function. s/SetupSandbox/setupSandbox

File: virtcontainers/cgroups.go [hunk 39]
```diff
@@ -32,6 +32,7 @@ const cgroupKataPath = "/kata/"
 // find kata containers cgroup path on host to prevent it
 // from grabbing the stats data.
 const cgroupKataPrefix = "kata"
+const cgroupSandboxPrefix = "kata-sandbox-"
```
[the remark this answers is not in the document]
Comment (egernst) [reply virtcontainers/cgroups.go]: I think we should use same prefix.  No need to be unique.

File: virtcontainers/container.go [hunk 40]
```diff
@@ -1422,8 +1428,8 @@ func (c *Container) detachDevices() error {
 	return nil
 }
 
-// creates a new cgroup and return the cgroups path
-func (c *Container) newCgroups() (err error) {
+// creates container cgroups in the host
```
Comment (egernst) [inline virtcontainers/container.go]:
```suggestion
// cgroupsCreate creates cgroups on the host for the associated container
```

File: virtcontainers/container.go [hunk 41]
```diff
@@ -1469,7 +1475,14 @@ func (c *Container) newCgroups() (err error) {
 	return nil
 }
 
-func (c *Container) deleteCgroups() error {
+// deletes container cgroups in the host
```
Comment (egernst) [inline virtcontainers/container.go]:
```suggestion
// cgroupsDelete deletes the cgroups on the host for the associated container
```

File: virtcontainers/container.go [hunk 42]
```diff
@@ -1497,13 +1510,19 @@ func (c *Container) deleteCgroups() error {
...
 	if err := cgroup.Delete(); err != nil {
-		return fmt.Errorf("Could not delete container cgroup %v: %v", c.state.CgroupPath, err)
+		return fmt.Errorf("Could not delete container cgroup path='%v': error='%v'", c.state.CgroupPath, err)
 	}
 
 	return nil
 }
 
-func (c *Container) updateCgroups(resources specs.LinuxResources) error {
+// updates container cgroups in the host
```
Comment (egernst) [inline virtcontainers/container.go]:
```suggestion
// cgroupsUpdate updates cgroups on the host for the associated container
```

File: virtcontainers/container.go [hunk 43]
```diff
@@ -1497,13 +1510,19 @@ func (c *Container) deleteCgroups() error {
...
+func (c *Container) cgroupsUpdate(resources specs.LinuxResources) error {
+
+	if c.state.CgroupPath == "" {
+		c.Logger().Debug("container has not cgroup path nothing to update")
```
Comment (egernst) [inline virtcontainers/container.go]:
```suggestion
		c.Logger().Debug("container does not have host cgroups: nothing to update")
```

File: virtcontainers/sandbox.go [hunk 44]
```diff
@@ -1872,3 +1880,264 @@ func (s *Sandbox) calculateSandboxCPUs() uint32 {
...
+			cpu.Mems += c.config.Resources.CPU.Mems + ","
+		}
+	}
+
+	cpu.Cpus = strings.Trim(cpu.Cpus, " \n\t,")
+
+	return validCPUResources(cpu)
+}
+
+// Finds the cgroups path for SandboxCgroupOnly config
```
Comment (egernst) [inline virtcontainers/sandbox.go]:
```suggestion
// setupSandboxCgroup creates sandbox cgroups for teh sandbox config
```

File: virtcontainers/sandbox.go [hunk 45]
```diff
@@ -1872,3 +1880,264 @@ func (s *Sandbox) calculateSandboxCPUs() uint32 {
...
+func (s *Sandbox) SetupSandboxCgroupOnly() error {
```
Comment (egernst) [inline virtcontainers/sandbox.go]:
```suggestion
func (s *Sandbox) setupSandboxCgroup() error {
```

File: virtcontainers/sandbox.go [hunk 46]
```diff
@@ -1872,3 +1880,264 @@ func (s *Sandbox) calculateSandboxCPUs() uint32 {
...
+
+	for _, cConfig := range s.config.Containers {
+		if cConfig.Annotations[annotations.ContainerTypeKey] == string(PodSandbox) {
+			PodSandboxConfig = &cConfig
+			break
+		}
+	}
+
+	if PodSandboxConfig == nil {
+		return fmt.Errorf("Failed to find cgroup path for Sandbox : Container of type '%s' not found", PodSandbox)
```
Comment (egernst) [inline virtcontainers/sandbox.go]: nit: remove space before ':'

File: virtcontainers/sandbox.go [hunk 47]
```diff
@@ -1872,3 +1880,264 @@ func (s *Sandbox) calculateSandboxCPUs() uint32 {
...
+	_, err := cgroupsNewFunc(cgroups.V1, cgroups.StaticPath(s.state.CgroupPath), &specs.LinuxResources{})
+	if err != nil {
+		return fmt.Errorf("Could not create sandbox cgroup in %v: %v", s.state.CgroupPath, err)
+
+	}
+
+	return nil
+}
+
+// Join to the Sandbox Cgroup, all the process the runtime creates after join are part of
```
Comment (egernst) [inline virtcontainers/sandbox.go]:
```suggestion
// joinSandboxCgroup adds the runtime PID to the sandbox defined in sandboxes' CgroupPath 
```

**Comment by egernst** [comment]
Thanks for the updates -- a few more comments.  Let me know when its ready for another round of review -- should be close! really looking forward to testing this with PodOverhead (code freeze is coming on 8/30)

**Comment by jcvenegas** [comment]
@egernst  please take a look to the last updates.

**Comment by egernst** [comment]
@bergwolf : the overhead is assumed to be just on host, we don’t size the VM any larger to account for it (today, anyway). So, the container sizing is the same as what is requested. Perhaps the default cpu/memory can be considered for this (recall that we start with a minimal cpu/memory, and then add in whatever is requested by the workload).
